### ISSUE TITLE
"Total fitness <= 0.0" error loading population state - memory issue or something else?

### ISSUE THREAD

**Issue by beaangelica** [MesserLab/SLiM#366]
Hi,
I am running a relatively simple model with a neutral burn-in phase and a second phase where I include a second mutation type drawn from a deleterious gamma DFE. Briefly, both models use 2000 individuals and 10 million sites on a single chromosome and record the effective population size as described in recipe 9.11 (with slight modifications). I am attaching my scripts at the end. However, I am seemingly having problems loading the population state from my burn-in into the second phase of my model.

My issue is this: 
I have run my burn-in phase and save the population state using outputFull() so that I have it in a separate .txt file. However, when I attempt to load the population state for the second phase (with the deleterious DFE), I get the error: 
"ERROR (Subpopulation::UpdateFitness): total fitness of all individuals is <= 0.0."

I do not get this error if I scale down the model and use a shorter chromosome (for both burn-in and second phase). Interestingly, I managed to get this (full-size) model to run once, but it was stopped after 600 generations (ca 5 hours) with an indiciation that it may have been stopped by the "out-of-memory handler" for the cluster I am running on. All other time I have gotten the same fitness <= 0.0 error every time after waiting for a few minutes for the population state file to load.

I have used output statements to isolate the cause to be something that happens either while or directly after loading the population state file, before the next generation can be run. The population state file is ca 4 GB and takes 5-7 minutes to load.

Is this an error that could be caused by a memory issue, or could something else be the problem? 

The models look as follows:

###### Burnin:
```
initialize() {
	initializeSLiMOptions(keepPedigrees=T);
	initializeMutationRate(2e-6);
	initializeMutationType("m1", 0.5, "f", 0.0);		// neutral
	initializeGenomicElementType("g1", m1, 1.0);
	initializeGenomicElement(g1, 0, 9999999);
	initializeRecombinationRate(1.2e-6);
}
1 late() {
	sim.addSubpop("p1", 2000);
	p1.setValue("previous_N", p1.individualCount);
	defineConstant("LOG", community.createLogFile("Log_Ne_Burnin_2Ki_10Ms_m6r6.csv"));
	LOG.addCycle();
	LOG.addCustomColumn("N(t-1)", "p1.getValue('previous_N');");
	LOG.addCustomColumn("N(t)", "p1.individualCount;");
	LOG.addCustomColumn("Ne_heterozygosity_Burnin", "estimateNe_Heterozygosity_all(p1);");
	LOG.addCustomColumn("Ne_inbreeding_Burnin", "estimateNe_Inbreeding(p1);");
}
2: late() {
	LOG.logRow();
	p1.setValue("previous_N", p1.individualCount);
}

40000 late() { 
	sim.outputFull("Ne_Burnin_2Ki_10Ms_m6r6__20Ng.txt");
	sim.simulationFinished();
}
```

###### DFE model:
```
initialize() {
	initializeSLiMOptions(keepPedigrees=T);
	initializeMutationRate(2e-6);
	initializeMutationType("m1", 0.5, "f", 0.0);		// neutral
	initializeMutationType("m2", 0.5, "g", -100, 0.1);	// DFE: mean -100, shape 0.1
	m2.convertToSubstitution = F;
	initializeGenomicElementType("g1",c(m1,m2), c(0.228,1.0));
	initializeGenomicElement(g1, 0, 9999999);
	initializeRecombinationRate(1.2e-6);
}

1 late() {
	sim.addSubpop("p1", 2000);
	sim.readFromPopulationFile(Ne_Burnin_2Ki_10Ms_m6r6__20Ng.txt);

	p1.setValue("previous_N", p1.individualCount);
	
	defineConstant("LOG", community.createLogFile("Log_Ne_DFE_2Ki_10Ms_m6r6.csv"));
	LOG.addCycle();
	LOG.addCustomColumn("N(t-1)", "p1.getValue('previous_N');");
	LOG.addCustomColumn("N(t)", "p1.individualCount;");
	LOG.addCustomColumn("Ne_4fold_heterozygosity_DFE", "estimateNe_Heterozygosity_fold(p1,m1);");
	LOG.addCustomColumn("Ne_0fold_heterozygosity_DFE", "estimateNe_Heterozygosity_fold(p1,m2);");
	LOG.addCustomColumn("Ne_all_heterozygosity_DFE", "estimateNe_Heterozygosity_all(p1);");
	LOG.addCustomColumn("Ne_inbreeding_DFE", "estimateNe_Inbreeding(p1);");
}

2: late() {
	LOG.logRow();
	p1.setValue("previous_N", p1.individualCount);
}

120000 late() {
	sim.outputFull("Ne_DFE_2Ki_10Ms_m6r6__40Ng.txt");
	p1.outputVCFSample(1000,replace=F,outputMultiallelics=F,filePath="DFE_1Kinds_40Ng.vcf");
	sim.outputFull();
	sim.simulationFinished();
}
```

###### Functions (included in both scripts above):
```
function (float)estimateNe_Heterozygosity_fold(o<Subpopulation>$ subpop, o<MutationType>$ mutType) {
	pi = calcHeterozygosity(subpop.genomes,sim.mutationsOfType(mutType));
	return pi / (4 * MU);
}

 function (float)estimateNe_Heterozygosity_all(o<Subpopulation>$ subpop) {
	pi = calcHeterozygosity(subpop.genomes);
	return pi / (4 * MU);
}

function (integer)tabulateFecundity(o<Subpopulation>$ subpop, i$ previous_N) {
	parentIDs = subpop.individuals.pedigreeParentIDs;
	rescaledParentIDs = parentIDs - min(parentIDs);
	return tabulate(rescaledParentIDs, previous_N - 1);
}

function (float)estimateNe_Inbreeding(o<Subpopulation>$ subpop) {
	previous_N = subpop.getValue("previous_N");
	k = tabulateFecundity(subpop, previous_N);
	return (previous_N * mean(k) - 2) / (mean(k) - 1 + var(k) / mean(k));
}
```

(I will add the burn-in file (4GB) to beaangelica/SLiM-dfe as soon as I figure out how to with GitLFS, but so far I have sent it to Ben for testing.)

Best regards, Bea Angelica Andersson

**Comment by bhaller** [comment]
Hi Bea!  Just to let you know, I'm not ignoring you but I just got diagnosed with COVID, and this issue seems a bit complicated, so it might wait until after I'm done with the brain fog and the low energy and everything.  :->

**Comment by bhaller** [comment]
Hi Bea.  Well, I tried to start looking at this but was blocked by this:

![constrictorTemp_789383](https://user-images.githubusercontent.com/10171101/208739466-71b7fc53-5ad3-43d6-ae8c-15238686ede0.png)

It looks like the .zip file you shared is corrupted in some way.  Maybe that "Sprend" platform messed with it.  Can you find a different way to upload it?  Also, if maybe you can generate a .gz file instead of a .zip file it might work better; worth a try.  Maybe you can share it through DropBox?

**Comment by beaangelica** [comment]
Hi Ben,

I have run some new tests to see if this issue is specific to our cluster or something general:
1) Yestarday I tried loading the population state file into another model that only included neutral mutations (i.e. only used the neutral m1 mutations and never introduced the m2 mutations that were drawn from a deleterious DFE). That model loaded the population state and ran fine for 500 generations (then I stopped it). When I included the m2 mutations, without changing anything else in the model, I got the same error about all individuals having a fitness <= 0.0.

2) I have then tried running models that have a chromosome size between 1K - 500K sites (instead of 10 million), using the same type of neutral burn-in as a first step and then loading the population state into a second step/run with a second mutation type added. These seem to work fine.

**Comment by bhaller** [comment]
Looks like the burn-in file has about 1.4 million segregating mutations, in 2000 diploid individuals, 4000 genomes.  That's somewhat large, but not unreasonably large, it seems to me.

One thing I'd note is that you might want to pass `binary=T` to `outputFull()` when you write to the file.  That will produce a binary file instead of a text file; it will be faster to save, smaller on disk, and faster to load.  Unless you actually want to see the information that's in the text file, binary is the way to go.

In the big picture, let me ask: given that your burn-in is pure neutral, and that it involves such a large number of mutations, is there a reason why you're not using tree-sequence recording?  I would guess that it would probably make your burn-in considerably faster, although I could be wrong.  Avoiding having to simulate large numbers of neutral mutations is a major motivation for the existence of tree-sequence recording.

But as to the actual issue here, of the error at the end of loading: I was unable to reproduce the bug.  It loaded fine for me, took some minutes and around 3 GB of memory but the load completed just fine and it didn't give me an error.  So then I wondered: is the problem actually the load at all, or is it what happens after the load in your model?  So I tried removing the load and the logging and such to make this bare-bones version of your model:

```
initialize() {
	initializeSLiMOptions(keepPedigrees=T);
	initializeMutationRate(2e-6);
	initializeMutationType("m1", 0.5, "f", 0.0);		// neutral
	initializeMutationType("m2", 0.5, "g", -100, 0.1);	// DFE: mean -100, shape 0.1
	m2.convertToSubstitution = F;
	initializeGenomicElementType("g1",c(m1,m2), c(0.228,1.0));
	initializeGenomicElement(g1, 0, 9999999);
	initializeRecombinationRate(1.2e-6);
}
1 late() {
	sim.addSubpop("p1", 2000);
}
120000 late() {
}
```

This results in the same error, often in the first generation.  The reason is the mutational dynamics you've defined.  With a mean of -100 for the gamma distribution, m2 mutations are generally going to be completely lethal; maybe occasionally they will be > 0.0, but usually they will be less.  (A quick test in EidosScribe shows that this distribution is negative about 98% of the time.)  Your mutation rate is 2e-6 and your chromosome is 1e7 bases long, so most of the time a new genome will receive at least one mutation, if not several.  And m2 mutations are the large majority of new mutations, 1.0 to 0.228.  What this amounts to is that when offspring are generated, it's pretty likely that every single one will be inviable due to having a lethal mutation.  When that happens – when every individual in the population has a fitness <= 0.0 – SLiM gives the error that you're seeing, because it is almost always indicative of a programming error, and also because it violates the assumptions of the Wright–Fisher model; the WF model assumes that it is always possible to draw parents to generate offspring, but if every individual's fitness is <= 0.0, drawing parents is impossible and so the script cannot be executed.

So, the loading is a red herring, and the problem is the mutational dynamics you've defined.  They drive the population extinct, but extinction is not allowed by the Wright–Fisher model (not just in SLiM, but in the original definition of it, too).  Once you fix that problem, I think things will work fine.  :->

By the way, in investigating this bug, I noticed a place in the loading code that is slower than it needs to be.  I will try tweaking that to speed it up, and load times might improve markedly as a result.  I'll post again on this issue once that has been pushed to GitHub, if it works.  But for now I'll close this issue, since I think your question has been resolved.  Good luck, and happy modeling!
